### PR TITLE
[Handshake] rewrite ConditionalBranchOp lowering with FIRRTL primitives

### PR DESCRIPTION
**Pull request by hanchenye**
Part of #178.

### CODE REVIEW

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 1]
```diff
@@ -758,57 +758,62 @@ bool HandshakeBuilder::visitHandshake(handshake::BranchOp op) {
...
-  rewriter.create<ConnectOp>(insertLoc, controlReady, control1ReadyOp);
+  // Connect ready signal of input and condition.
+  auto selectedResultReady = rewriter.create<MuxPrimOp>(
+      insertLoc, trueResultReady.getType(), conditionData, trueResultReady,
+      falseResultReady);
+
+  rewriter.create<ConnectOp>(
+      insertLoc, argReady,
+      rewriter.create<AndPrimOp>(insertLoc, selectedResultReady.getType(),
+                                 selectedResultReady, conditionValid));
```
Comment (mikeurbach) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:810]: Should the `argReady` signal depend only on `selectedResultReady` and `conditionValid`? What if the condition is valid, but the arg is not valid yet? I think in this case, the arg would be marked ready when it shouldn't, but I may be missing something.

Comment (mikeurbach) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:810]: Related, I think the logic `selectedResultReady & conditionArgValid` could be used for both `argReady` and `conditionReady`. Does that make sense? I'm looking at Figure 5 [here](http://www.cs.columbia.edu/~sedwards/papers/edwards2019compositional.pdf). I think that would address my correctness concern as well as simplify this be removing one `AndPrimOp` from the current implementation.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 2]
```diff
@@ -758,57 +758,62 @@ bool HandshakeBuilder::visitHandshake(handshake::BranchOp op) {
...
 
-  if (!op.isControl()) {
-    Value argData = argSubfields[2];
-    Value result0Data = result0Subfields[2];
-    rewriter.create<ConnectOp>(insertLoc, result0Data, argData);
-  }
+  auto conditionArgValid = rewriter.create<AndPrimOp>(
+      insertLoc, conditionValid.getType(), conditionValid, argValid);
+
+  auto conditionNot = rewriter.create<NotPrimOp>(
```
Comment (mikeurbach) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:779]: Nice, I didn't realize `ConditionalBranchOp` only ever has two outputs. That simplifies things a bit. I guess if we ever need to generalize it to variadic outputs, we can use decoders or something along those lines.

**Review by mikeurbach — COMMENTED** [review]
@hanchenye this looks pretty good. I do have one concern below.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp (hunk 1 above)
Comment (hanchenye) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp]: Make sense. I've updated the implementation.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp (hunk 2 above)
Comment (hanchenye) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:779]: Exactly!

**Review by mikeurbach — APPROVED** [review]
Looks good to me!

**Pull request merged**